### DISCUSSION TITLE
Binary distributions

### DISCUSSION THREAD

**Discussion by planetf1** [open-quantum-safe/liboqs#1625] [General]
Is there a list of binary distributions (particularly aided by the core team, or at least well-supported) of liboqs? Obviously there are more built by individuals for their own needs.

I can build liboqs from source (and have the most flexibility), but when explaining to others how to use liboqs, pointing to a easily accessible built version - such as installed from their standard linux distro, or a common package manager (like homebrew on macos) can make things easier & less error prone

Looking around I found a few references:

* https://packages.fedoraproject.org/pkgs/liboqs/ - looks to be current. I've myself installed liboqs from fedora 39. This is really useful (also builds at https://koji.fedoraproject.org/koji/packageinfo?packageID=37370 ) . I presume this is the current official initiative for fedora?
* https://copr.fedorainfracloud.org/coprs/rrelyea/liboqs/ - I presume this was an early, now defunct initiative. It targets a broader range of distributions than the above
* https://code.launchpad.net/ubuntu/+source/liboqs - an earlier .8.x version for a range of ubuntu versions. Non-official?
* https://repology.org/project/liboqs/information - some additional random hits
* https://formulae.brew.sh/formula/liboqs - homebrew. This looks current, presumably official?
* https://aur.archlinux.org/packages?O=0&SeB=nd&K=liboqs&outdated=&SB=p&SO=d&PP=50&submit=Go - archlinux. unofficial?
* https://www.freshports.org/security/liboqs - FreeBSD (0.8.x)

No doubt there are many more

Is this documented anywhere? Would it be useful to add those with some backing/confidence that are being maintained (happy to volunteer helping out in doc etc, search in more detail & catalog)

Logically the topic could continue with the other subprojects like python binding.

Many thanks!

**Comment by beldmit** [comment]
I am the maintainer of liboqs and oqs provider in Fedora, I confirm this is the official Fedora initiative. But we intentionally build only NIST-approved algorithms

**Comment by baentsch** [comment]
> Is there a list of binary distributions (particularly aided by the core team, or at least well-supported) of liboqs?

Not really. OQS never really cared about binary distributions for lack of (human and CI) resources (and experiences, say with non-Linux/non-x64 platforms). Even a general platform support strategy is currently only "in the making" (see https://github.com/open-quantum-safe/liboqs/pull/1605), so anyone doing distros (cudos to @beldmit ) _may_ face "surprises". 

Only `oqsprovider` does [provide binaries as part of its releases](https://github.com/open-quantum-safe/oqs-provider/releases/tag/0.5.2) as the maintainer of that project considers it a key feature for a "binary loadable module".

> happy to volunteer helping out in doc etc, search in more detail & catalog

Please, by all means, go ahead! It would also be nice if you could collect (CI) environments we could rely on if we decide to support specific platforms. It would help [M1](https://github.com/open-quantum-safe/liboqs/pull/1605#discussion_r1413009379) as well as not really well-supported arcane [IBM platforms](https://github.com/open-quantum-safe/liboqs/pull/1605#discussion_r1414062858) (sorry, couldn't resist :).

**Comment by planetf1** [comment]
thanks.. makes sense to only distribute the NIST candidates in built form for now. 
Sorry for the slow reply (many threads) - I'll do some more digging ... and be back here.

**Reply by baentsch** [reply]
Ping... Two months on, any news on the above? 

Otherwise, my suggestion is to close the discussion assuming lack of interest and leave it to interested (non-OQS) parties to create/maintain installation packages independently: "Ready-to-run code" in my eyes is a property of a "commercially backed" project and hence should be motivated/driven by commercial entities (for the platforms important to them, e.g., RHEL/Fedora as per the comment above), not volunteer contributors (that may focus on the wrong platforms or rather prioritize improving the functionality or properties of the code).

Finally, if there's still interest on your side in/willingness to shape the future direction of the project, please consider providing feedback to things like [this](https://github.com/orgs/open-quantum-safe/discussions/1689) or [this](https://github.com/open-quantum-safe/tsc/issues/2).

**Reply by raybeecham** [reply]
Thank you for the update; I appreciate the efforts to refine OQS towards production readiness and am interested in contributing feedback or assistance to help navigate this transition effectively. While I'm not entirely sure if I possess the knowledge or the skills to assist I'd like to see what I can help with.

**Reply by ametzler** [reply]
Hello,
Debian is keeping liboqs out of  stable releases (and therefore the testing distribution) https://bugs.debian.org/1000303 with the rationale
```
The upstream wishes to keep the package out of stable distributions for
now, until the NIST Post-Quantum Cryptography standardization project
reaches its conclusion. For the time being the package may be used for
prototyping and evaluation in unstable.
```
Does this still reflect _current_ wishes of liboqs upstream, now after the NIST project was finalized?
TIA, cu Andreas

**Reply by SWilson4** [reply]
> Does this still reflect _current_ wishes of liboqs upstream, now after the NIST project was finalized? TIA, cu Andreas

At the moment, I would say "yes"—we don't yet have an implementation of [the final version of FIPS 204](https://github.com/open-quantum-safe/liboqs/issues/1891). After that lands, perhaps we can revisit this decision. @open-quantum-safe/tsc Might this be worth discussing in tomorrow's meeting?

**Reply by baentsch** [reply]
As @SWilson4 pointed out, OQS at this point in time does not include software in line with the final versions of all published NIST standards (see https://github.com/open-quantum-safe/liboqs/issues/1894#issuecomment-2456309867 for details). There further are several unresolved discussions as to what it takes to transform the software into a state that it can be considered safe for use by "Average Joe", e.g., https://github.com/open-quantum-safe/tsc/issues/1.

The mission of the project _currently_ is unchanged to "provide software for evaluation and prototyping", so I'd second the "Yes" by @SWilson4.

That said, @ametzler , you may want to build your own opinion as to whether it's OK to include OQS in stable releases. I am a more cautious person than the committee majority of corporate America controlling OQS with the position of ["just drop usage warnings" and "let the user decide"](https://github.com/PQCA/TAC/issues/44). I am of the opinion that this is security software, warranting additional caution/quality concerns. This opinion also is based on my insight of the state of the software and its origins -- but may be tainted by my position of a maintainer (with a thus perceived responsibility to provide reliable software).

> @open-quantum-safe/tsc Might this be worth discussing in tomorrow's meeting?

Yes, a verbal exchange of thoughts between people with time to participate only via meetings and those contributing publicly/via GH surely may be beneficial to understand the mutual concerns and goals.

**Reply by dstebila** [reply]
> > @open-quantum-safe/tsc Might this be worth discussing in tomorrow's meeting?
> 
> Yes, a verbal exchange of thoughts between people with time to participate only via meetings and those contributing publicly/via GH surely may be beneficial to understand the mutual concerns and goals.

I don't think that's an appropriate response. We have meetings to facilitate these discussions; the TSC will not be the sole location for those discussions, but it can be one venue for them.

**Reply by baentsch** [reply]
> I don't think that's an appropriate response. We have meetings to facilitate these discussions; the TSC will not be the sole location for those discussions, but it can be one venue for them.

I didn't think I "said" anything other than that, sorry if that got across in an unintended way: My goal is to ensure  people are not disadvantaged due to their (lack of) mastery of language or (un)availability in specific time zones (or general lack of time) by having different venues for discussion (incl. meetings and GH).

**Reply by ametzler** [reply]
Thanks for the quick responses, continuing to keep this in Debian/unstable only seems to be the right thing to do.

**Comment by adrien-n** [comment]
Hi,

(apologies if an earlier version of this message got through your notifications or inbox :) )

I'm looking at updating liboqs and the openssl provider in Ubuntu and Debian. There are at least two aspects: a) update packages, regardless of whether they are shipped in a distribution release, b) offer them to users.

Ubuntu releases an LTS version every 2 years and an interim ones every 6 months. There are planning stages months before each development cycle and feature freezes a couple month before the release. My current plan is to prepare an update of liboqs for 25.04 which is an interim version and hope that you are satisfied with liboqs' status by 26.04 LTS so that users can benefit from it.

I discussed that with other Ubuntu/Canonical developers and we feel like there is no perfect and clear-cut solution. We want to include oqs early enough, but not too early. Except we don't have crystal balls either. The best we can do is prepare everything, do the integration, and either not do the final integration steps for now, or do it but hide it behind a feature flag. One thing we all agreed upon is that only hybrid cryptography makes sense and we'll make sure to make this clear in Ubuntu. Moreover, nobody felt that it is possible to commit to support for post-quantum cryptography software today: our motivation is to help make things move forward.

My rationale for targetting the next interim release (25.04) is how much work will still have to happen afterwards in the distribution, and if we aim for wider availability in 26.04, all of it will have to be done within a year or so. If not 26.04, the following Ubuntu release that will be used as widely will 28.04 and I think everyone will agree it's getting too far in the future (especially since it means real world deployments at least several months later).

There is no intention to actively promote these packages and as I said above, I'd like to offer them in the distribution so that people can start evaluating them and provide feedback. An additional benefit will be the CI that runs on amd64/arm64/armhf/ppc64el/s390x (not riscv sadly due to its slowness).

What do you think about this? Does it seem to sound to begin preparations today but gate everything so that actually enabling requires an active operation from users?

**Reply by baentsch** [reply]
Thanks for this thorough description of events (at least I wasn't aware of until now), @adrien-n ! Just this statement triggers a question in my mind:

> One thing we all agreed upon is that only hybrid cryptography makes sense and we'll make sure to make this clear in Ubuntu

Does this mean you will want to disable "plain PQ" algorithms or keep them but just attached with a documented warning? If the former, do you expect a config option to do this or would you patch the code accordingly? In the prototype-only oqsprovider all of this is configurable (exactly for experimentation as per its stated usage goal) but we didn't consider adding an option to do this (yet) in OpenSSL (which for now is not based on oqs): See https://github.com/openssl/openssl/pull/25848 . Worthwhile opening a separate discussion there?

**Reply by adrien-n** [reply]
That's a good question. I don't have a full answer at the moment however. Ideally, there wouldn't be any specific configuration or patch: following upstream's choices should always come first. Specific options and patches in distributions should only exist to solve distribution problems IMO.

I think that in practice, the first step will be an updated package and then some integration. It's possible that for Ubuntu it will be limited to a PPA and/or we use warnings (or use runtime configuration to disable everything?). Some of that will depend on openssl and gnutls for instance (if oqs is needed while building gnutls, then a PPA won't be appropriate: an openssl provider is more flexible than gnutls' typical approach AFAIU). Later on we'll probably either expand what's enabled, or restrict it: it's difficult to say for now. :) 

I've had bad experiences with the past where different runtimes couldn't communicate together as expected because they were configured slightly differently. I'm a bit scared that each distribution comes up with its own set of algorithms with only partial overlap. We can discuss that further for openssl indeed.

NB: with current openssl and ubuntu policies, Ubuntu will probably have to stick to the next openssl LTS (has to be 3.5 or maybe 3.6) for four years after the LTS release.

**Reply by baentsch** [reply]
Thanks for that timeline heads-up @adrien-n. fwiw, my hope is that OpenSSL will have good, OQS-independent PQC implementations available for all NIST-standardized algorithms in such LTS (at least I try to help make that happen, but wouldn't give a guarantee at this time, particularly considering to the many options "around" those algorithms wrt key formats or the various "hybridization options"... Should those still not have "stabilized"/finalized by the time of your LTS, then an availability of `oqsprovider` in "unstable" may be the best approach to hedge all bets).

Concerning the question above (hybrid only support or all-option support), I'd personally very much prefer the latter: [IANA has assigned code points to "pure" ML-KEM](https://www.iana.org/assignments/tls-parameters/tls-parameters.xhtml#tls-parameters-8), so why not give users the option to use them? The crypto code for that exists anyway -- albeit I agree, it has not stood the test of time...

**Reply by dstebila** [reply]
@adrien-n Just getting caught up on some older issues. Did you get the information you need to be able to proceed? Or is there more you're looking for from us?

**Reply by adrien-n** [reply]
@baentsch It's difficult to predict future state and I wouldn't worry that much about it. Releasing a distribution version is taking many many late decisions for plenty of packages and typically it's not possible to take these decisions earlier on. I'd say that's the life of releasing a distribution. ;p 

Regarding support for hybrid-only or pure PQC, I think we all want to enable pure but the unknown is exactly when that will happen and whether it's done in several steps (especially wrt shipping vs. enabling by default). We definitely want to start with hybrid so pure will be something in addition to and not instead of. As such, for shipping, it's probably mostly going to be a matter of time and release schedule deadlines. For enabling by default, it's another matter but things have ample time to change by then, and it's not impossible to enable features post-release if needed.


@dstebila I've investigated more one of the first issue for the Debian packaging which is also probably the least interesting: licenses and copyrights. Debian packaging requires that every file has license and copyright data. Unfortunately the data in liboqs' upstreams can be lost during updates because it's tied to file paths rather than stored in the files themselves. This also makes it difficult to automatically created the corresponding database. I went for a restricted set of algorithms and opened the following PRs:

- https://github.com/pq-crystals/kyber/pull/93
- https://github.com/pq-crystals/dilithium/pull/99
- https://github.com/sphincs/sphincsplus/pull/64
- https://github.com/PQClean/PQClean/pull/579

I decided to work on the file upstreams as I didn't have to track their origins and the bet is that I selected the relevant ones (mostly at least: I know it's definitely not perfect but the work should be transferable). As you can see, I didn't get feedback for these PRs, which is not a surprise because it's definitely non-exciting stuff.
Is there a good way to integrate these changes in liboqs even if these upstreams or upstreams-squared didn't pickt hem up? I didn't yet have time to do it myself and see the result on automation but I shall do that this week.

**Comment by dstebila** [comment]
@planetf1 In the 2024-12-05 OQS TSC meeting, we discussed this issue. The list of resources you started to collect is great. Would you be willing to turn this into a table, with contact information (Github handles or otherwise) for each distribution? Also there's the question of whether we would want to put this in the Wiki tab of Github or somewhere else.

**Reply by planetf1** [reply]
sure, we can start to collate, just need to figure out where.  I'd presume the obvious place would be somewhere in the https://github.com/open-quantum-safe/www repo so it appears on https://openquantumsafe.org/ . Not sure where it fits the best. Could be under liboqs, but it may be broader in scope (for example oqsprovider ?)

**Reply by dstebila** [reply]
Okay, how about starting a new page `distributions` under [`applications`](https://github.com/open-quantum-safe/www/tree/main/applications)?

**Reply by dstebila** [reply]
Hi Nigel, do you have any cycles to kick this off?

**Reply by baentsch** [reply]
Would this do for now?

 [![Packaging status](https://repology.org/badge/vertical-allrepos/liboqs.svg)](https://repology.org/project/liboqs/versions)

 https://repology.org/project/liboqs/versions

**Reply by dstebila** [reply]
Yes, that's cool! Thanks for finding it. Where to do you think is a good place for us to link to it? I've let https://openquantumsafe.org/dashboard.html get a little out of date but that would be one option.

**Reply by baentsch** [reply]
I'd agree, that would be a reasonable location as it's AFAIK not publicly linked (right?) and thus would inform only ourselves as per the original purpose stated at the beginning of this discussion.

Putting this onto other, more publicly visible location would make me uneasy in turn as it might imo be as misleading and counterproductive as any other "badge": The two key disadvantages I'd see: 1) It will further reduce the incentive for anyone, incl. PQCA companies to actually contribute to make the code factually fit for wider use ("it's already in distros, why contribute to improve?"). 2) If OQS would more widely publish this, this can further mis-lead users to assume OQS considers the code to be ready for "prime time" (while in reality this is just indicative of other people packaging the library, without making assertions/improvements about its utility/quality).

That said, we of course cannot stop the PQCA marketing guys to use this in their communications to create further pressure on OQS to deliver and maintain (presumably) "productive code" for their customers :-(

Anyway, whatever you do, _please_ do not publish any of this for `oqsprovider`: That code is currently so far behind any sensible goals for reliable use (standards, secure coding practices, good design, issues resolution, ...) that I would want to avoid _any_ further misconception regarding that project.

**Comment by dstebila** [comment]
As discussed in the 2025-03-03 TSC meeting, we'd still like to have a page listing points of contact for distributions using OQS so that we can reach out to them more easily if need be. @planetf1 any ability to help with this?

**Comment by defanator** [comment]
Hi everyone. I'm currently working on packaged distributions of liboqs and oqs-provider for a few internal projects, but as a part of this work I came up with some bits that would help to produce .debs (unique for every target distro we use) from helpers kept in base source distributions of both liboqs and oqs-provider.

I know that the question "where to keep packaging harness - alongside with the base code or separately" could be controversial, and over the years I've been using different approaches for different software and code bases; nevertheless, I wanted to ask whether these projects would accept PRs for making general packaging tasks easier to accomplish.

Thanks!

PS: we started from .debs that could be produced by CPack directly from current code base of liboqs, and here's output of lintian(1) from one of these packages:

```
E: liboqs: extended-description-is-empty
E: liboqs: lacks-ldconfig-trigger usr/lib/liboqs.so.0.15.0
E: liboqs: malformed-contact Maintainer www.openquantumsafe.org
E: liboqs: no-changelog usr/share/doc/liboqs/changelog.gz (native package)
E: liboqs: no-copyright-file
E: liboqs: no-shlibs usr/lib/liboqs.so.0.15.0
E: liboqs: unstripped-binary-or-object [usr/lib/liboqs.so.0.15.0]
W: liboqs: link-to-shared-library-in-wrong-package usr/lib/liboqs.so.0.15.0 [usr/lib/liboqs.so]
W: liboqs: package-name-doesnt-match-sonames liboqs9
W: liboqs: pkg-config-unavailable-for-cross-compilation [usr/lib/pkgconfig/liboqs.pc]
W: liboqs: synopsis-too-long
```

Not a big deal for testing purposes, but definitely could be improved with adapting e.g. debuild(1) from devscripts. Also, current cmake/cpack instructions are not suitable for producing e.g. .rpms or Alpine's .apks (not sure it can do the latter at all).

**Reply by baentsch** [reply]
Thanks for touching base @defanator . I personally have decided to shelve the aspiration to make available binary packages (see https://github.com/open-quantum-safe/oqs-provider/issues/94) for two prime reasons: 1) Such "simple to deploy" mechanisms are going squarely against the stated ["do not use this code productively"](https://github.com/open-quantum-safe/oqs-provider/?tab=readme-ov-file#component-disclaimer) warnings in the project that are there for _very_ good reasons and 2) There's simply not enough support ("contributorship") within OQS to deal with the incremental load that ready-made binaries would cause in the user base (e.g., a substantial increase in possible "problem finders" or more security exposure the project isn't geared to handle).

**Reply by defanator** [reply]
These are fair points - thanks for getting back here @baentsch.

re: "very good reasons" for existing disclaimers around liboqs/oqs-provider production readiness - we are still trying to assess these in our particular use cases; if you have any particular issues/discussions raised here in the past, I'd appreciate some pointers.
